jax/_src/numpy/ufuncs.py

```diff
@@ -2181,6 +2182,15 @@ def ldexp(x1: ArrayLike, x2: ArrayLike, /) -> Array:
   return _where(isinf(x1) | isnan(x1) | (x1 == 0), x1, x)
 
 
+@ldexp.defjvp
+def _ldexp_jvp(x2, primals, tangents):
+  (x1,) = primals
+  (x1_dot,) = tangents
+  out = ldexp(x1, x2)
+  out_dot = exp2(x2) * x1_dot
```

Comment:
Would it be better to write this as `out_dot = ldexp(x1_dot, x2)`?

Reply:
@jakevdp Thought about that too, but it gives me the following error:

<details>

```python3
(venv) $ pytest -rP -rx -n auto tests/lax_numpy_test.py -k testGradLdexp1
=============================== test session starts ================================
platform darwin -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0
rootdir: /Users/carlos/Desktop/jax
configfile: pyproject.toml
plugins: hypothesis-6.100.0, nengo-4.0.0, xdist-3.5.0
16 workers [1 item]       
F                                                                            [100%]
===================================== FAILURES =====================================
__________________________ NumpyGradTests.testGradLdexp1 ___________________________
[gw0] darwin -- Python 3.12.3 /Users/carlos/venv/bin/python3

>   check_grads(lambda x: jnp.ldexp(x, n), (x,), 1)

tests/lax_numpy_test.py:6147: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jax/_src/public_test_util.py:310: in check_grads
    _check_grads(f, args, order)
jax/_src/public_test_util.py:297: in _check_grads
    _check_jvp(f, partial(api.jvp, f), args, err_msg=fwd_msg)
jax/_src/public_test_util.py:235: in check_jvp
    v_out, t_out = f_jvp(args, tangent)
tests/lax_numpy_test.py:6147: in <lambda>
    check_grads(lambda x: jnp.ldexp(x, n), (x,), 1)
jax/_src/numpy/ufuncs.py:2189: in _ldexp_jvp
    out = ldexp(x1, x2)
jax/_src/numpy/ufuncs.py:2157: in ldexp
    x, e = _normalize_float(x1)
jax/_src/numpy/ufuncs.py:2130: in _normalize_float
    x1 = _where(cond, x * _lax_const(x, 1 << info.nmant), x)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   return lax.select(condition, x_arr, y_arr) if not is_always_empty else x_arr
E   jax._src.source_info_util.JaxStackTraceBeforeTransformation: AssertionError
E   
E   The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.
E   
E   --------------------

jax/_src/numpy/util.py:442: JaxStackTraceBeforeTransformation

The above exception was the direct cause of the following exception:

self = <lax_numpy_test.NumpyGradTests testMethod=testGradLdexp1>, n = -3
dtype = <class 'jax.numpy.float32'>

    @jtu.sample_product(
      n=range(-4, 5),
      dtype=[jnp.float32, jnp.float64],
    )
    def testGradLdexp(self, n, dtype):
      rng = jtu.rand_default(self.rng())
      x = rng((), dtype)
>     check_grads(lambda x: jnp.ldexp(x, n), (x,), 1)

tests/lax_numpy_test.py:6147: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
jax/_src/public_test_util.py:310: in check_grads
    _check_grads(f, args, order)
jax/_src/public_test_util.py:303: in _check_grads
    _check_vjp(f, partial(api.vjp, f), args, err_msg=rev_msg)
jax/_src/public_test_util.py:260: in check_vjp
    cotangent_out = conj(vjpfun(conj(cotangent)))
jax/_src/tree_util.py:477: in __call__
    return self.fun(*args, **kw)
jax/_src/api.py:1880: in _vjp_pullback_wrapper
    ans = fun(*args)
jax/_src/tree_util.py:477: in __call__
    return self.fun(*args, **kw)
jax/_src/interpreters/ad.py:149: in unbound_vjp
    arg_cts = backward_pass(jaxpr, True, consts, dummy_args, cts)
jax/_src/interpreters/ad.py:237: in backward_pass
    cts_out = reducing_transposes[eqn.primitive](
jax/_src/pjit.py:2310: in _pjit_transpose
    transpose_jaxpr, attrs_tracked = _pjit_transpose_trace(
jax/_src/linear_util.py:352: in memoized_fun
    ans = call(fun, *args)
jax/_src/pjit.py:2282: in _pjit_transpose_trace
    transpose_jaxpr, _, consts, attrs_tracked = pe.trace_to_jaxpr_dynamic(
jax/_src/profiler.py:333: in wrapper
    return func(*args, **kwargs)
jax/_src/interpreters/partial_eval.py:2272: in trace_to_jaxpr_dynamic
    jaxpr, out_avals, consts, attrs_tracked = trace_to_subjaxpr_dynamic(
jax/_src/interpreters/partial_eval.py:2295: in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers_)
jax/_src/linear_util.py:193: in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
jax/_src/interpreters/ad.py:251: in closed_backward_pass
    return backward_pass(jaxpr.jaxpr, transform_stack, jaxpr.consts,
jax/_src/interpreters/ad.py:237: in backward_pass
    cts_out = reducing_transposes[eqn.primitive](
jax/_src/pjit.py:2310: in _pjit_transpose
    transpose_jaxpr, attrs_tracked = _pjit_transpose_trace(
jax/_src/linear_util.py:352: in memoized_fun
    ans = call(fun, *args)
jax/_src/pjit.py:2282: in _pjit_transpose_trace
    transpose_jaxpr, _, consts, attrs_tracked = pe.trace_to_jaxpr_dynamic(
jax/_src/profiler.py:333: in wrapper
    return func(*args, **kwargs)
jax/_src/interpreters/partial_eval.py:2272: in trace_to_jaxpr_dynamic
    jaxpr, out_avals, consts, attrs_tracked = trace_to_subjaxpr_dynamic(
jax/_src/interpreters/partial_eval.py:2295: in trace_to_subjaxpr_dynamic
    ans = fun.call_wrapped(*in_tracers_)
jax/_src/linear_util.py:193: in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
jax/_src/interpreters/ad.py:251: in closed_backward_pass
    return backward_pass(jaxpr.jaxpr, transform_stack, jaxpr.consts,
jax/_src/interpreters/ad.py:240: in backward_pass
    cts_out = get_primitive_transpose(eqn.primitive)(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

t = Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=2/0)>
which = UndefinedPrimal(ShapedArray(bool[]))
cases = (UndefinedPrimal(ShapedArray(float32[])), UndefinedPrimal(ShapedArray(float32[])))

    def _select_transpose_rule(t, which, *cases):
>     assert not ad.is_undefined_primal(which)
E     AssertionError

jax/_src/lax/lax.py:4286: AssertionError
================================ 1 failed in 2.86s =================================
(venv) $ 
```

</details>

Looks like a transpose rule needs to be defined? (Possibly related: #9129.)

Reply:
Yeah, if the JVP of a primitive depends on itself, it is no longer linear by definition and needs a transpose rule.

That seems like the right thing though: if `ldexp` is the right way to implement `x * 2 ** y`, then surely it's also the right way to implement `x_dot * 2 ** y`?

Reply:
There's not (yet) a `custom_transpose` operation, but you can indirectly define it using `custom_vjp`

Reply:
After thinking about it more, I think I agree with your comment:

> `ldexp` represents the mathematical operation $x * 2^y$ <ins>with platform-specific implementation details when custom kernels are available</ins>. In this case, the optimal implementation in JAX would be to write `x * 2 ** y`, which would have the side effect of making `custom_jvp` unnecessary.

Is it alright if we go with that?

Reply:
Yeah, let's go for it!

We should still require an integer exponent just in case we want to dispatch to a more specific kernel in the future.

Reply:
Updated accordingly.